src/ngraph/op/batch_norm.cpp

```diff
@@ -21,51 +21,98 @@
+                                                   std::shared_ptr<ngraph::Node> beta,
+                                                   std::shared_ptr<ngraph::Node> input,
+                                                   std::shared_ptr<ngraph::Node> mean,
+                                                   std::shared_ptr<ngraph::Node> variance)
+    : BatchNormBase(
+          "BatchNormInference", eps, check_single_output_args({gamma, beta, input, mean, variance}))
 {
-    m_bn_input_shape = get_input_shape(INPUT);
-    NODE_VALIDATION_ASSERT(this, m_bn_input_shape.size() >= 2)
-        << "Input argument must have rank of at least 2 (input argument shape: " << m_bn_input_shape
-        << ").";
+    constructor_validate_and_infer_types();
+}
 
-    NODE_VALIDATION_ASSERT(this, m_bn_input_shape[1] != 0)
-        << "Input argument's channel dimension must have size of at least 1 (input argument shape: "
-        << m_bn_input_shape << ").";
+ngraph::op::BatchNormTraining::BatchNormTraining(double eps,
+                                                 std::shared_ptr<ngraph::Node> gamma,
+                                                 std::shared_ptr<ngraph::Node> beta,
+                                                 std::shared_ptr<ngraph::Node> input)
+    : BatchNormBase("BatchNormTraining", eps, check_single_output_args({gamma, beta, input}))
+{
+    constructor_validate_and_infer_types();
+}
 
-    auto& et = get_input_element_type(INPUT);
+ngraph::op::BatchNormTraining::BatchNormTraining(double eps,
+                                                 std::shared_ptr<ngraph::Node> gamma,
+                                                 std::shared_ptr<ngraph::Node> beta,
+                                                 std::shared_ptr<ngraph::Node> input,
+                                                 std::shared_ptr<ngraph::Node> mean,
+                                                 std::shared_ptr<ngraph::Node> variance)
+    : BatchNormBase(
+          "BatchNormTraining", eps, check_single_output_args({gamma, beta, input, mean, variance}))
+{
+    constructor_validate_and_infer_types();
+}
+
+void ngraph::op::BatchNormInference::validate_and_infer_types()
+{
+    auto bn_input_shape = get_input_shape(INPUT);
+    BatchNormBase::validate_and_infer_types();
     auto in_size = get_input_size();
+    auto& et = get_input_element_type(INPUT);
+    NODE_VALIDATION_ASSERT(this, in_size == 5)
+        << "Argument count must be 5 (received argument count: " << in_size << ").";
+    set_output_size(1);
+    set_output_type(0, et, bn_input_shape);
+}
 
+void ngraph::op::BatchNormTraining::validate_and_infer_types()
+{
+    auto bn_input_shape = get_input_shape(INPUT);
+    BatchNormBase::validate_and_infer_types();
+    auto in_size = get_input_size();
+    auto& et = get_input_element_type(INPUT);
     NODE_VALIDATION_ASSERT(this, in_size == 3 || in_size == 5)
```

Comment:
I think we should split the 3-input and 5-input variants into separate ops.

Reply:
Yes. One is used in training, the other in inference.